app/src/org/commcare/android/database/global/GlobalDatabaseUpgrader.java

```
@@ -78,4 +91,40 @@ private boolean upgradeTwoThree(SQLiteDatabase db) {
             db.endTransaction();
         }
     }
+
+    /**
+     * Prior to multiple application seating, the FormsProvider and the InstanceProvider were both
+     * using one global database for all forms/instances. Now that we can have multiple apps
+     * installed at once, we need to have one forms db and one instances db per app. This method
+     * performs the necessary one-time migration from a global db that still exists on a device
+     * to the new per-app system
+     */
+    private boolean upgradeProviderDb(SQLiteDatabase db, ProviderUtils.ProviderType type) {
+        File oldDbFile = CommCareApplication._().getDatabasePath(type.getOldDbName());
+        if (oldDbFile.exists()) {
+            File newDbFile = CommCareApplication._().getDatabasePath(
+                    ProviderUtils.getProviderDbName(type, getInstalledAppRecord(c, db).getApplicationId()));
+            if (!oldDbFile.renameTo(newDbFile)) {
+                // Big problem, should potentially crash here ?
```

Should this fail harder in some way?